File: LeetCode/538-convert-bst-to-greater-tree/ex_538.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* convertBST(TreeNode* root) {
        sum = 0;
        return helper(root);
    }
private:
    int sum;
    TreeNode* helper(TreeNode* node) {
        if (node == NULL)   return NULL;
        helper(node->right);
        sum += node->val;
        node->val = sum;
        helper(node->left);
        return node;
    }
};

File: LeetCode/538-convert-bst-to-greater-tree/ex_538.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    s = 0
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        self.s = 0
        return self.helper(root)
        
    def helper(self, node):
        if not node:    return
        self.helper(node.right)
        self.s += node.val
        node.val = self.s
        self.helper(node.left)
        return node